src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -194,20 +194,91 @@ public override int GetHashCode()
     /// </summary>
     public class CallGraph
     {
-        // TODO:
-        // This is the method that should be invoked to verify cycles of interest,
-        // i.e. where each callable in the cycle is type parametrized.
-        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined.
-        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        /// <summary>
+        /// Given a cycle of call graph edges, determines if the cycle is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
+        /// Returns true if the cycle is valid, false if invalid.
+        /// </summary>
+        internal static bool VerifyCycle(params CallGraphEdge[] edges)
```

Comment:
How does this relate to the diagnostic QS6213: "The given expression constrains the type parameter(s) 'a"? Will `VerifyCycle` be used to generate compile errors for the test cases?

Reply:
I am hoping to separate out the logic that throws Diagnostic from the tooling for the QDK as much as possible. When this is put to use, we will see to what extent that can be accomplished and to what degree this code needs to throw Diagnostics.

Reply:
My question is mainly that if these test cases are invalid, why do they compile?

Reply:
That's because the call graph and its features are not fully integrated into the language yet. 🙂 That will have to be another PR, or more likely a set of PR's as I think that will take quite a bit of work.

Reply:
OK, in the mean time will `VerifyCycle` be used for anything?

Reply:
There is nothing in this PR that uses the VerifyCycle other than the unit tests.

Reply:
I meant: what was the reason for writing `VerifyCycle`? Is it only to generate diagnostics at some point in the future, or was there another use case?

Reply:
Once we can integrate this with monomorphization, diagnostics will be generated based on this logic. I'm thinking in the precondition of the monomorphization rewrite step. So If there are invalid cycles in the call graph, we can throw an error diagnostic at that point.

Reply:
I think the rewrite step precondition would not be ideal, since diagnostics generated during the precondition won't show up automatically in the IDE (only when running `dotnet build`) - so the errors would not be noticed while typing. (Unless #544 is resolved.)

Reply:
Okay, we can discuss that in more detail when the time comes.